Extract logic for get/put node selection

gavin-norman-sociomantic opened the pull request:
(Based on #32.)

nemanja-boric-sociomantic commented on src/dhtproto/client/internal/NodeHashRanges.d:
```diff
@@ -57,6 +59,115 @@ public final class NodeHashRanges : NodeHashRangesBase
+        TODO: test the logic for retrying the request on other nodes which
+        previously covered the hash. This will require a full neo implementation
+        of the Redistribute request. See
+        https://github.com/sociomantic/dhtnode/issues/624
```
non-tsunami link.

gavin-norman-sociomantic replied:
:+1: I'll create a new issue in tsunami.

nemanja-boric-sociomantic commented on src/dhtproto/client/internal/NodeHashRanges.d:
```diff
@@ -57,6 +59,115 @@ public final class NodeHashRanges : NodeHashRangesBase
+    {
+        auto nodes = this.getNodesForHash(h, node_hash_ranges);
+        if ( nodes.length == 0 )
+            return true;
```
Why not make this return code inverse? If there's no record, return `false`, that seems more logical.

gavin-norman-sociomantic replied:
I know what you mean. The most common usage requires a `!` then, but I guess that doesn't really matter.

nemanja-boric-sociomantic commented on src/dhtproto/client/internal/NodeHashRanges.d:
```diff
@@ -57,6 +59,115 @@ public final class NodeHashRanges : NodeHashRangesBase
+        if ( nodes.length == 0 )
+            return true;
+
+        foreach_reverse ( node_hash_range; nodes )
```
This `foreach_reverse` indicates that the ordering is somehow important. Maybe it should be added to the method's documentation for the parameter `getNodesForHash`.

gavin-norman-sociomantic replied:
It already is described...?

>During a data redistribution, more than one node may be responsible
for a given key. In this case, the node that was most recently
reported as being responsible for the key is queried first, followed
by others (in order) until the record is located, an error occurs,
or no node has the record.

gavin-norman-sociomantic replied:
Oh I see what you mean. `getNodesForHash` doesn't mention how the returned list is ordered.

gavin-norman-sociomantic replied:
Weird, it appears it's actually unsorted! (I think this is what the TODO above is referring to ;)

nemanja-boric-sociomantic commented on src/dhtproto/client/internal/NodeHashRanges.d:
```diff
@@ -57,6 +59,115 @@ public final class NodeHashRanges : NodeHashRangesBase
+                break;
+        }
+
+        return false;
```
This fixed `return false` doesn't seem to be covering the case where we get the list of nodes responsible for the key, but they all return `true`. Perhaps rewrite this as:

```
+            // If we got the record or an error occurred, don't try more nodes
 +            if ( !try_next_node )
 +                return true;
 +        }
 +
 +        return false;
```

gavin-norman-sociomantic replied:
The caller can already detect this case (they are the ones who decide whether on the return value of get), so it doesn't need to be specifically covered here.

gavin-norman-sociomantic replied:
Actually, the return value isn't needed at all. The caller can detect all possible cases externally.

nemanja-boric-sociomantic commented on src/dhtproto/client/internal/NodeHashRanges.d:
```diff
@@ -57,6 +59,115 @@ public final class NodeHashRanges : NodeHashRangesBase
+                break;
+        }
+
+        return false;
```
This fixed `return false` seems to ignore the case where we tried all nodes that we got as the responsible, but nobody contains the record anymore. Is that plausible scenario to keep yourself from?

gavin-norman-sociomantic replied:
The caller can already detect this case (they are the ones who decide whether on the return value of `get`), so it doesn't need to be specifically covered here.

nemanja-boric-sociomantic commented on src/dhtproto/client/internal/NodeHashRanges.d:
```diff
@@ -57,6 +59,115 @@ public final class NodeHashRanges : NodeHashRangesBase
+        ulong newest_order;
+        foreach ( i, node_hash_range; nodes )
+        {
+            if ( node_hash_range.order > newest_order )
```
To confirm: `node_hash_range.order` starts from `1`? Maybe it's safer to do `if (node_hash_range.order >= newest_order)`.

gavin-norman-sociomantic replied:
There should never be two nodes with the same `order`. I can add checks that verify this.

nemanja-boric-sociomantic replied:
No, it was my brain that stuck yesterday. This piece of code is actually fine.

Commit:
Add comment explaining use of bitswap in unittest

Commit:
Sort list returned by NodeHashRangesBase.getNodesForHash

Code that is already using this method assumes that the returned list of
node hash ranges is sorted by their `order` field. The code to actually do
this sorting was missing. This commit adds it, along with a unittest to
check the sorting, and updates the code that uses this method.

Commit:
Extract logic for get/put node selection

Previously this logic was only required in the Get / Put requests so was
situated in the modules with the implementations of those requests. Now
another request will need this logic, so it is extracted into a shared
location.

gavin-norman-sociomantic commented:
Updated.

nemanja-boric-sociomantic commented on src/dhtproto/client/internal/NodeHashRanges.d:
```diff
@@ -198,6 +201,14 @@ private class NodeHashRangesBase
 
+        bool sortPred ( NodeHashRange e1, NodeHashRange e2 )
+        {
+            assert(e1.order != e2.order);
```
Too early for `verify`?

gavin-norman-sociomantic replied:
Mmm... Is it released yet?

nemanja-boric-sociomantic replied:
Yes, in ocean `v2.10.0`/`v3.4.0`.

gavin-norman-sociomantic replied:
Cool, I can update.

nemanja-boric-sociomantic replied:
> Cool, I can update.

I assume you figured out that you can't update, since we don't support this ocean yet?

gavin-norman-sociomantic replied:
Oh I forgot about this. What do you mean we don't support it yet?

nemanja-boric-sociomantic replied:
If dhtproto/swarm is using ocean prior to `v3.4.0`, `verify` is not yet available.

gavin-norman-sociomantic replied:
I could update the submodule in this (semver-free) branch.

gavin-norman-sociomantic replied:
I won't bother. It introduces zillions of deprecation warnings. Something for a separate PR.

nemanja-boric-sociomantic commented on src/dhtproto/client/internal/NodeHashRanges.d:
```diff
@@ -57,6 +59,94 @@ public final class NodeHashRanges : NodeHashRangesBase
+
+    ***************************************************************************/
+
+    public void getFromNode ( hash_t h, ref NodeHashRange[] node_hash_ranges,
```
Yeah, without return value what this method is doing (and how it should be used) is much clearer.

gavin-norman-sociomantic commented:
Rebased on `neo`. Unblocked.
